### PR TITLE
Validate namespaceConfig with the apps that sharing the target namespace

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward giantswarm/giantswarm#16209

My original plan was to check the target namespace object but I realized app-admission-controller can't fetch any objects from workload clusters. So I check app CRs with the same target namespace instead.

### CODE REVIEW

File: pkg/key/app.go [hunk 1]
```diff
@@ -27,6 +27,10 @@ func AppName(customResource v1alpha1.App) string {
 	return customResource.Spec.Name
 }
 
+func AppNamespace(customResource v1alpha1.App) string {
```
Comment (tomahawk28) [inline pkg/key/app.go:30]: Adding this AppNamespace for the sake of simplicity.

File: pkg/validation/app.go [hunk 2]
```diff
@@ -103,6 +108,59 @@ func (v *Validator) validateConfig(ctx context.Context, cr v1alpha1.App) error {
...
+
+	for _, app := range apps {
+		if key.AppNamespace(cr) != key.AppNamespace(app) {
+			continue
+		}
+
+		targetAnnotations := key.AppNamespaceAnnotations(app)
+		if targetAnnotations != nil && annotations != nil {
+			for k, v := range targetAnnotations {
+				originalValue, ok := annotations[k]
```
Comment (tomahawk28) [inline pkg/validation/app.go:141]: Check only if another app has the same label key.

File: pkg/validation/app.go [hunk 3]
```diff
@@ -103,6 +108,59 @@ func (v *Validator) validateConfig(ctx context.Context, cr v1alpha1.App) error {
...
+				if ok && originalValue != v {
+					return microerror.Maskf(validationError, "app %#q is modifying annotation %#q in the target namespace %#q to the different value",
```
Comment (rossf7) [inline pkg/validation/app.go:143]: Current message is OK but I think we could add more detail to help the user. 

I'd like to know which app CR has the same label and maybe also what the current value is.

**Comment by tomahawk28** [comment]
I changed it as below. 

```
app `kiam` is modifying labels `monitoring` in the target namespace `kube-system` from value `enabled` to `disabled`, collide to app `another-kiam-1`
```

WDYT?

File: pkg/validation/app.go [hunk 4]
```diff
@@ -103,6 +108,59 @@ func (v *Validator) validateConfig(ctx context.Context, cr v1alpha1.App) error {
...
+			continue
+		}
+
+		targetAnnotations := key.AppNamespaceAnnotations(app)
+		if targetAnnotations != nil && annotations != nil {
+			for k, v := range targetAnnotations {
+				originalValue, ok := annotations[k]
+				if ok && originalValue != v {
+					return microerror.Maskf(validationError, "app %#q is modifying annotation %#q in the target namespace %#q from value %#q to %#q, collide to app %#q",
+						key.AppName(cr), k, key.AppNamespace(cr), originalValue, v, app.Name)
```
Comment (rossf7) [inline pkg/validation/app.go:144]:
```suggestion
					return microerror.Maskf(validationError, "app %#q annotation %#q for target namespace %#q collides with value %#q for app %#q",
						key.AppName(cr), k, key.AppNamespace(cr), v, app.Name)
```

This is a tricky one. I tried to make the message shorter. Also the user knows which value they set so I think we can leave that out.

This is easier to review in the unit test message IMO.

File: pkg/validation/app.go [hunk 5]
```diff
@@ -103,6 +108,59 @@ func (v *Validator) validateConfig(ctx context.Context, cr v1alpha1.App) error {
...
+			}
+		}
+
+		targetLabels := key.AppNamespaceLabels(app)
+		if targetLabels != nil && labels != nil {
+			for k, v := range targetLabels {
+				originalValue, ok := labels[k]
+				if ok && originalValue != v {
+					return microerror.Maskf(validationError, "app %#q is modifying labels %#q in the target namespace %#q from value %#q to %#q, collide to app %#q",
+						key.AppName(cr), k, key.AppNamespace(cr), originalValue, v, app.Name)
```
Comment (rossf7) [inline pkg/validation/app.go:155]:
```suggestion
						return microerror.Maskf(validationError, "app %#q label %#q for target namespace %#q collides with value %#q for app %#q",
						key.AppName(cr), k, key.AppNamespace(cr), v, app.Name)
```

File: pkg/validation/app_test.go [hunk 6]
```diff
@@ -735,6 +735,174 @@ func Test_ValidateMetadataConstraints(t *testing.T) {
...
+						NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+							Annotations: map[string]string{
+								"linkerd.io/inject": "disabled",
+							},
+						},
+						Version: "1.3.0-rc1",
+					},
+				},
+			},
+			expectedErr: "app `kiam` is modifying annotation `linkerd.io/inject` in the target namespace `kube-system` from value `enabled` to `disabled`, collide to app `another-kiam-1`",
```
Comment (rossf7) [inline pkg/validation/app_test.go:824]:
```suggestion
			expectedErr: "app `kiam` annotation `linkerd.io/inject` for target namespace `kube-system` collides with value `disabled` for app `another-kiam-1`",
```

File: pkg/validation/app_test.go [hunk 7]
```diff
@@ -735,6 +735,174 @@ func Test_ValidateMetadataConstraints(t *testing.T) {
...
+						NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+							Labels: map[string]string{
+								"monitoring": "disabled",
+							},
+						},
+						Version: "1.3.0-rc1",
+					},
+				},
+			},
+			expectedErr: "app `kiam` is modifying labels `monitoring` in the target namespace `kube-system` from value `enabled` to `disabled`, collide to app `another-kiam-1`",
```
Comment (rossf7) [inline pkg/validation/app_test.go:864]:
```suggestion
			expectedErr: "app `kiam` label `monitoring` for target namespace `kube-system` collides with value `disabled` from app `another-kiam-1`",
```

File: pkg/validation/app.go (hunk 4 above)
Comment (tomahawk28) [reply pkg/validation/app.go:144]: This is much better, Thanks!

File: pkg/validation/app_test.go (hunk 6 above)
Comment (tomahawk28) [reply pkg/validation/app_test.go:824]: Thanks, actually I copy the result from my unittest 😄

**Review by rossf7 — APPROVED** [review]
LGTM 🚀

**Pull request merged**